### PR TITLE
DNS Proxy emits error monitor events & proxy metrics

### PR DESCRIPTION
**Pull request by raybejjani**
Processing a DNS request requires a number of steps that may fail. We
now emit error events, and send a REFUSED response when this occurs.
This is more in-line with our other proxies.
Note: This doesn't emit an event when the packet parsing fails. That is
handled by miekg/dns and it always just drops the request and lets it
time out.

The DNS proxy now also emits metrics like other proxies.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6498)

### CODE REVIEW

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

File: daemon/daemon.go [hunk 1]
```diff
@@ -1422,26 +1423,32 @@ func (d *Daemon) bootstrapFQDN() (err error) {
...
 				}
 			}
 			if ep == nil { // this is a hard fail. We cannot proceed
 				err := fmt.Errorf("Cannot find matching endpoint for IPs %s or %s", srcAddr, dstAddr)
 				log.WithError(err).Error("cannot find matching endpoint")
 				return err
 			}
+			dstPort, err := strconv.Atoi(dstPortStr)
+			if err != nil {
+				log.WithError(err).WithField(logfields.Port, dstPortStr).Error("cannot parse destination port")
```
Comment (aanm) [inline daemon/daemon.go]: @raybejjani Should we continue if we can't parse the destination port?

Comment (raybejjani) [reply daemon/daemon.go]: "Yes". It is only used for reporting and I treated it as a non-fatal error. I'm not sure if there is a better way to express this situation, though. I'll add comments to these cases anyway, though.
Some of the other errors, where we do return, might also be "non-fatal". Some seem to panic later, and others not (e.g. `ep == nil` would panic the e.Log function).  Do you see a different way to structure this function? I would love if it was more clear what was reporting (metrics and Accesslog/monitor events) and what was business logic around updating the DNS cache info.

Comment (raybejjani) [reply daemon/daemon.go]: @aanm I added a bunch of comments. Let me know if it's more clear now, or if I should try something else.

**Review by aanm — APPROVED** [review]

**Comment by raybejjani** [comment]
@aanm I had to change the code a bit, since the metrics relies on dstPort, but we extracted that later in the function. The diff (of code, some comments are reformatted) to what you approved is below, and it's in the final `pkg/fqdn: DNS Proxy emits metrics` commit. Sorry!

```diff
diff --git a/daemon/daemon.go b/daemon/daemon.go
index 032bd7363..8d348c077 100644
--- a/daemon/daemon.go
+++ b/daemon/daemon.go
@@ -1428,17 +1432,21 @@ func (d *Daemon) bootstrapFQDN() (err error) {
                                }
                        }

-                       var dstIPStr, dstPortStr string
-                       if ep == nil {
-                               dstIPStr, dstPortStr, err = net.SplitHostPort(dstAddr)
-                               if err != nil {
-                                       // This may be fatal, but we handle that case below, if ep == nil
-                                       log.WithError(err).Error("cannot extract endpoint IP from DNS request")
-                               } else {
-                                       // ep needs to be non-nil for the access log. One of the ends of this connection will be an endpoint.
-                                       ep = endpointmanager.LookupIPv4(dstIPStr)
+                       var dstPort int
+                       dstIPStr, dstPortStr, err := net.SplitHostPort(dstAddr)
+                       if err != nil {
+                               // This may be fatal, but we handle that case below, if ep == nil
+                               log.WithError(err).Error("cannot extract endpoint IP from DNS request")
+                       } else {
+                               if dstPort, err = strconv.Atoi(dstPortStr); err != nil {
+                                       log.WithError(err).WithField(logfields.Port, dstPortStr).Error("cannot parse destination port")
                                }
                        }
+
+                       if ep == nil {
+                               // ep needs to be non-nil for the access log. One of the ends of this connection will be an endpoint.
+                               ep = endpointmanager.LookupIPv4(dstIPStr)
+                       }
                        if ep == nil {
                                // This is a hard fail. We cannot proceed because record.Log requires a
                                // non-nil ep, and we also don't want to insert this data into the
@@ -1450,10 +1458,6 @@ func (d *Daemon) bootstrapFQDN() (err error) {
                                ep.UpdateProxyStatistics("dns", uint16(dstPort), ingress, !ingress, accesslog.VerdictError)
                                return err
                        }
-                       dstPort, err := strconv.Atoi(dstPortStr)
-                       if err != nil {
-                               log.WithError(err).WithField(logfields.Port, dstPortStr).Error("cannot parse destination port")
-                       }

                        qname, responseIPs, TTL, CNAMEs, err := dnsproxy.ExtractMsgDetails(msg)
                        if err != nil {
```

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
This is so weird... I can't reproduce this locally but it always fails `
k8s-1.8.K8sChaosTest Restart with long lived connections TCP connection is not dropped when cilium restarts` in CI

**Comment by raybejjani** [comment]
test-me-please (https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/8797/flowGraphTable/)

**Comment by eloycoto** [comment]
test-me-please

**Pull request merged**